src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+                          .ToImmutableArray();
+
+            // Now try to generate all the members that will go in the new class. This is a bit
+            // circular.  In order to generate some of the members, we need to know about the type.
+            // But in order to create the type, we need the members.  To address this we do two
+            // passes. First, we create an empty version of the class.  This can then be used to
+            // help create members like Equals/GetHashCode.  Then, once we have all the members we
+            // create the final type.
+            var namedTypeWithoutMembers = CreateNamedType(className, capturedTypeParameters, default);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var constructor = CreateConstructor(compilation, className, properties, generator);
+
+            // Generate Equals/GetHashCode.  Only readonly properties are suitable for these
+            // methods.  We can defer to our existing language service for this so that we
+            // generate the same Equals/GetHashCode that our other IDE features generate.
+            var propertiesForEqualsAndGetHashCode = ImmutableArray<ISymbol>.CastUp(
+                properties.WhereAsArray(p => p.SetMethod == null));
+
+            var equalsAndGetHashCodeService = document.GetLanguageService<IGenerateEqualsAndGetHashCodeService>();
+                        
+            var equalsMethod = await equalsAndGetHashCodeService.GenerateEqualsMethodAsync(
+                document, namedTypeWithoutMembers, propertiesForEqualsAndGetHashCode, 
+                localNameOpt: ICodeDefinitionFactoryExtensions.OtherName, cancellationToken).ConfigureAwait(false);
+            var getHashCodeMethod = await equalsAndGetHashCodeService.GenerateGetHashCodeMethodAsync(
+                document, namedTypeWithoutMembers,
+                propertiesForEqualsAndGetHashCode, cancellationToken).ConfigureAwait(false);
+
+            var members = ArrayBuilder<ISymbol>.GetInstance();
+            members.AddRange(properties);
+            members.Add(constructor);
+            members.Add(equalsMethod);
+            members.Add(getHashCodeMethod);
+
+            var namedTypeSymbol = CreateNamedType(className, capturedTypeParameters, members.ToImmutableAndFree());
+            return namedTypeSymbol;
+        }
+
+        private static INamedTypeSymbol CreateNamedType(
+            string className, ImmutableArray<ITypeParameterSymbol> capturedTypeParameters, ImmutableArray<ISymbol> members)
+        {
+            return CodeGenerationSymbolFactory.CreateNamedTypeSymbol(
+                attributes: default, Accessibility.Internal, modifiers: default, 
+                TypeKind.Class, className, capturedTypeParameters, members: members);
+        }
+
+        private (ImmutableArray<IPropertySymbol>, ImmutableDictionary<IPropertySymbol, string>) GenerateProperties(INamedTypeSymbol anonymousType)
```

Comment:
>(ImmutableArray<IPropertySymbol>, ImmutableDictionary<IPropertySymbol, string>) [](start = 16, length = 79)

Nit: although it makes the line long, I'd name the tuple elements, as documentation in source and QuickInfo.

Reply:
will do.,